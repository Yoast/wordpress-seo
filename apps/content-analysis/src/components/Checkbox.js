import React from "react";
import PropTypes from "prop-types";
import { isFunction, noop } from "lodash-es";

class Checkbox extends React.PureComponent {
	/**
	 * Initializes the Checkbox component
	 *
	 * @param {Object}   props             The component's props.
	 * @param {string}   props.id          The id of the checkbox.
	 * @param {string}   props.value       The value of the checkbox.
	 * @param {string}   props.label       The label in front of the checkbox.
	 * @param {function} props.onChange    Callback to receive the new text.
	 *
	 * @returns {void}
	 */
	constructor( props ) {
		super( props );

		this.handleChange = this.handleChange.bind( this );
	}

	/**
	 * Calls the onChange prop with the changed text.
	 *
	 * @param {Event} event The checkbox change event.
	 *
	 * @returns {void}
	 */
	handleChange( event ) {
		const { onChange } = this.props;

		if ( isFunction( onChange ) ) {
			onChange( event.target.checked );
		}
	}

	/**
	 * Renders the Checkbox component.
	 *
	 * @returns {void}
	 */
	render() {
		const { id, value, label } = this.props;

		return (
			<React.Fragment>
				<label htmlFor={ id }>{ label }</label>
				<input
					type="checkbox"
					id={ id }
					name={ id }
					checked={ value }
					onChange={ this.handleChange }
				/>
			</React.Fragment>
		);
	}
}

Checkbox.propTypes = {
	id: PropTypes.string.isRequired,
	value: PropTypes.bool.isRequired,
	label: PropTypes.string,
	onChange: PropTypes.func,
};

Checkbox.defaultProps = {
	label: "",
	onChange: noop,
};

export default Checkbox;
