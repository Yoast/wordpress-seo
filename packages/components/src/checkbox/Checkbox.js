import PropTypes from "prop-types";
import React, { useCallback } from "react";
import { FieldGroup } from "../field-group";
import "./checkbox.css";

/**
 * Represents the input type checkbox with a label, inside a FieldGroup.
 *
 * @param {Object} props The component properties.
 *
 * @returns {JSX.Element} A React component that wraps around the HTML checkbox.
 */
export default function Checkbox( props ) {
	const handleChange = useCallback( ( event ) => {
		props.onChange( event.target.value );
	}, [ props.onChange ] );

	return (
		<FieldGroup wrapperClassName="yoast-field-group yoast-field-group__checkbox">
			<input type="checkbox" id={ props.id } checked={ props.checked } onChange={ handleChange } />
			<label htmlFor={ props.id }>{ props.label }</label>
		</FieldGroup>
	);
}

Checkbox.propTypes = {
	id: PropTypes.string.isRequired,
	label: PropTypes.oneOfType( [
		PropTypes.string,
		PropTypes.arrayOf( PropTypes.node ),
		PropTypes.node,
	] ).isRequired,
	checked: PropTypes.bool,
	onChange: PropTypes.func.isRequired,
};

Checkbox.defaultProps = {
	checked: false,
};
