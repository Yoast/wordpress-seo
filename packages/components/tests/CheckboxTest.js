import React from "react";
import renderer from "react-test-renderer";

import Checkbox from "../src/checkbox/Checkbox";

describe( "Checkbox", () => {
	it( "matches the snapshot", () => {
		const component = renderer.create(
			<Checkbox id="test-id" label="test label" />
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );

	it( "matches the snapshot when an array is provided as a label", () => {
		const component = renderer.create(
			<Checkbox id="test-id" label={ [ "test label ", "using arrays" ] } />
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );

	it( "executes callback once", () => {
		const event = {
			target: {
				checked: false,
			},
		};
		const onChange = jest.fn();
		const component = renderer.create(
			<Checkbox
				id="testCallback"
				onChange={ onChange }
				label="testCallbackLabel"
			/>
		);

		const tree = component.toTree();
		tree.props.onChange( event );

		expect( onChange ).toHaveBeenCalledTimes( 1 );
	} );
} );
