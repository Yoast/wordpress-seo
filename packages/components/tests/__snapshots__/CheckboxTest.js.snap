// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox matches the snapshot 1`] = `
<div
  className="yoast-field-group yoast-field-group__checkbox"
>
  <input
    checked={false}
    id="test-id"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="test-id"
  >
    test label
  </label>
</div>
`;

exports[`Checkbox matches the snapshot when an array is provided as a label 1`] = `
<div
  className="yoast-field-group yoast-field-group__checkbox"
>
  <input
    checked={false}
    id="test-id"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="test-id"
  >
    test label 
    using arrays
  </label>
</div>
`;
